providers/luadns/api.go

```diff
@@ -0,0 +1,250 @@
+	return nil
+}
+
+func (l *luadnsProvider) fetchDomainList() error {
+	l.domainIndex = map[string]uint32{}
+	var bodyString, err = l.get("/zones", "GET", requestParams{})
+	if err != nil {
+		return fmt.Errorf("failed fetching domain list from LuaDNS: %s", err)
+	}
+	var dr zoneResponse
+	json.Unmarshal(bodyString, &dr)
+	for _, domain := range dr {
+		l.domainIndex[domain.Name] = domain.ID
+	}
+	return nil
+}
+
+func (l *luadnsProvider) getDomainID(name string) (uint32, error) {
+	if l.domainIndex == nil {
+		if err := l.fetchDomainList(); err != nil {
+			return 0, err
+		}
+	}
+	id, ok := l.domainIndex[name]
+	if !ok {
+		return 0, fmt.Errorf("'%s' not a zone in luadns account", name)
+	}
+	return id, nil
+}
+
+func (l *luadnsProvider) createDomain(domain string) error {
+	params := jsonRequestParams{
+		"name": domain,
+	}
+	if _, err := l.get("/zones", "POST", params); err != nil {
+		return fmt.Errorf("failed create domain (LuaDNS): %s", err)
+	}
+	return nil
+}
+
+func (l *luadnsProvider) createRecord(domainID uint32, rec jsonRequestParams) error {
+	if _, err := l.get(fmt.Sprintf("/zones/%d/records", domainID), "POST", rec); err != nil {
+		return fmt.Errorf("failed create record (LuaDNS): %s", err)
+	}
+	return nil
+}
+
+func (l *luadnsProvider) deleteRecord(domainID uint32, recordID uint32) error {
+	if _, err := l.get(fmt.Sprintf("/zones/%d/records/%d", domainID, recordID), "DELETE", requestParams{}); err != nil {
+		return fmt.Errorf("failed delete record (LuaDNS): %s", err)
+	}
+	return nil
+}
+
+func (l *luadnsProvider) modifyRecord(domainID uint32, recordID uint32, rec jsonRequestParams) error {
+	if _, err := l.get(fmt.Sprintf("/zones/%d/records/%d", domainID, recordID), "PUT", rec); err != nil {
+		return fmt.Errorf("failed update (LuaDNS): %s", err)
+	}
+	return nil
+}
+
+func (l *luadnsProvider) getRecords(domainID uint32) ([]domainRecord, error) {
+	var bodyString, err = l.get(fmt.Sprintf("/zones/%d/records", domainID), "GET", requestParams{})
+	if err != nil {
+		return nil, fmt.Errorf("failed fetching record list from LuaDNS: %s", err)
+	}
+	var dr recordResponse
+	json.Unmarshal(bodyString, &dr)
+	var records []domainRecord
+	for _, rec := range dr {
+		if rec.Type == "SOA" {
+			continue
+		}
+		records = append(records, rec)
+	}
+	return records, nil
+}
+
+func (l *luadnsProvider) get(endpoint string, method string, params any) ([]byte, error) {
+	client := &http.Client{}
+	var req, err = l.makeRequest(endpoint, method, params)
+	if err != nil {
+		return []byte{}, err
+	}
+	req.Header.Set("Accept", "application/json")
+	req.SetBasicAuth(l.creds.email, l.creds.apikey)
+	// LuaDNS has a rate limit of 1200 request per 5 minute.
+	// So we do a very primitive rate limiting here - delay every request for 250ms - so max. 4 requests/second.
+	time.Sleep(250 * time.Millisecond)
+	resp, err := client.Do(req)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	bodyString, _ := io.ReadAll(resp.Body)
+	if resp.StatusCode == 200 {
```

Comment:
Move `if resp.StatusCode == 200` closer to where `resp` is set. (after the `if err != nil` check)